analyses/cell-type-wilms-tumor-14/scripts/utils/01_anchor_transfer_seurat_functions.R

```diff
@@ -0,0 +1,107 @@
+library(dplyr)
+library(Seurat)
+library(ggpubr)
+library(zellkonverter)
+library(SingleCellExperiment)
+
+
+run_anchorTrans <- function(path_anal, scratch_out_dir, results_out_dir,
+                            ref_obj, sample, 
+                            level = "celltype", # celltype or compartment
+                            k_weight = 50,
+                            unknown_cutoff = 0.5, ndims = 20) {
+  # perspective output files
+  filename <- file.path(results_out_dir, paste0(sample, "_", level,".pdf"))
+  filename_csv <- file.path(results_out_dir, paste0(sample, "_", level,".csv"))
+
+  
+  sample_obj <- SeuratObject::LoadSeuratRds( file.path(path_anal,"scratch","00_preprocessing_rds",paste0(sample,".rdsSeurat")) )
+
+  
+  # set row names as gene symbol, since reference obj uses gene symbol by default
+  # length(intersect(rownames(ref_obj), rownames(sample_obj))) # make sure gene symbol consistency
+  # length(intersect(VariableFeatures(ref_obj), rownames(sample_obj))) # candidate genes used for transfer
+  # unique <- sample_obj[["RNA"]]@meta.data$gene_ids[!duplicated(sample_obj[["RNA"]]@meta.data$gene_symbol)]
+  # sample_obj <- sample_obj[unique,]
+  # rownames(sample_obj) <- sample_obj[["RNA"]]@meta.data$gene_symbol
+  
+  # find anchors
+  anchors <- FindTransferAnchors(reference = ref_obj, query = sample_obj, dims = 1:ndims)
+  nanchors <- nrow(anchors@anchors)
+  # clean annotation, too few stroma, immune and endo
+  # ref_obj@meta.data <- ref_obj@meta.data %>%
+  #   mutate(annot = case_when(compartment == "stroma" ~ "stroma",
+  #                            compartment == "immune" ~ "immune",
+  #                            compartment == "endothelium" ~ "endothelium",
+  #                            TRUE ~ celltype))
+  if (level == "compartment") {
+    ref_obj@meta.data <- ref_obj@meta.data %>%
+      mutate(annot = compartment)
+  } else {
+    ref_obj@meta.data <- ref_obj@meta.data %>%
+      mutate(annot = celltype)
+  }
+  ref_obj@meta.data$annot <- factor(ref_obj@meta.data$annot)
+  # transfer labels
+  predictions <- TransferData(anchorset = anchors, refdata = ref_obj$annot, dims = 1:ndims, k.weight = k_weight)
+  predictions <- mutate(predictions, predicted.id = case_when(prediction.score.max < unknown_cutoff ~ "Unknown",
+                                                              TRUE ~ predicted.id))
+  sample_obj <- AddMetaData(object = sample_obj, metadata = predictions)
+  
+  # save prediction table
+  write.csv(predictions, file = filename_csv)
+  
+  # anchor transfer plots
+  plot_anchorTrans(path_anal = path_anal, 
+                   scratch_out_dir = scratch_out_dir, 
+                   results_out_dir = results_out_dir,
```

Comment:
Do you maybe want to put these in the plots directory? I know that to do this you would need to add a line or lines to `.gitignore`, as they are too large to commit, but that might be worth doing for organization.